\documentclass{article}
\usepackage[french,english]{babel}
\usepackage[utf8]{inputenc}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{listings}
\usepackage{xcolor}
\lstset{frame=single,backgroundcolor=\color{pink}}
\newcommand{\sge}{\href{http://gridscheduler.sourceforge.net}{\em{SGE}}}
\newcommand{\git}{\href{https://github.com/}{\em{git}}}
\newcommand{\qmake}{\href{http://gridscheduler.sourceforge.net/htmlman/htmlman1/qmake.html}{\em{qmake}}}
\newcommand{\make}{\href{http://www.gnu.org/software/make/manual/make.html}{\em{make}}}
\newcommand{\velocity}{\href{http://velocity.apache.org/}{\em{velocity}}}
\newcommand{\bwa}{\href{http:/todo.org/}{\em{bwa}}}
\newcommand{\VCF}{VCF}
\newcommand{\redspan}[1]{\textcolor{red}{#1}}

\author{Pierre Lindenbaum PhD\\\href{https://twitter.com/yokofakun}{@yokofakun}\\ \href{mailto:plindenbaum@yahoo.fr}{pierre.lindenbaum@univ-nantes.fr}\\ \url{http://plindenbaum.blogspot.com}\\Institut du Thorax. Nantes. France}
\title{xml4ngs}
\date{}
\begin{document}
\maketitle

\section{Introduction}

Xml4ngs defines a XML schema describing a Exome/Whole genome sequencing result. This schema
is used to generate a Makefile-based workflow of analysis.

\section{Overview}
\begin{center}
\includegraphics[scale=0.4]{workflow02.png}
\end{center}

\subsection{About MAKE}
GNU \make{}  is a "utility that automatically builds executable programs and libraries from source code by reading files called makefiles which specify how to derive the target program. Though integrated development environments and language-specific compiler features can also be used to manage a build process".\\
A Makefile that contains lists of dependencies between different source files and object files. It also contains lists of commands that should be executed to satisfy these dependencies. \make{} uses the timestamps of files and the information of the files' existence to automate rebuilding of applications (targets in make) as well as the rules that we specify in the Makefile.\\
A complex Makefile is generated by xml4ngs in order to analyze a NGS experiment.

\subsection{About qmake}
Qmake is a parallel, distributed \make{} utility. Scheduling of the parallel make tasks is done by \sge{}. It is based on gmake (GNU \make{}).
\subsection{About Apache Velocity}
The Apache \velocity{} Engine is a free open-source templating engine. \velocity{} is used in xml4ngs to generate a complex Makefile describing the tasks to analyze a NGS experiment.
\begin{center}
\includegraphics[scale=0.6]{workflow05.png}
\end{center}


\section{Install and Compile}

\subsection{Requirements}
The tools require the following packages/softwares to be compiled:
\begin{enumerate}
\item SUN JDK 7 . \url{http://www.oracle.com/technetwork/java/index.html}.
\item apache ANT .\url{http://ant.apache.org/}. Ant will use \href{http://ant.apache.org/ivy/}{IVY} to fetch the library for apache \velocity{}.
\end{enumerate}

Pull xml4ngs from github: \url{https://github.com/lindenb/xml4ngs}
Edit the file build.properties if needed.

type 'ant' to generate the tools.
\begin{lstlisting}[language=bash]
$ ant


Buildfile: /home/lindenb/src/xml4ngs/build.xml


generate-sources:
     [exec] parsing a schema...
     [exec] compiling a schema...
     [exec] com/github/lindenb/xml4ngs/FastqType.java
     [exec] com/github/lindenb/xml4ngs/LanesType.java
     [exec] com/github/lindenb/xml4ngs/ObjectFactory.java
     [exec] com/github/lindenb/xml4ngs/PairType.java
     [exec] com/github/lindenb/xml4ngs/ProjectType.java
     [exec] com/github/lindenb/xml4ngs/PropertiesType.java
     [exec] com/github/lindenb/xml4ngs/PropertyType.java
     [exec] com/github/lindenb/xml4ngs/SampleType.java
     [exec] com/github/lindenb/xml4ngs/SequencesType.java

dist/ilmn2project.jar:
    [mkdir] Created dir: /home/lindenb/src/xml4ngs/tmp
    [javac] Compiling 1 source file to /home/lindenb/src/xml4ngs/tmp
      [jar] Building jar: /home/lindenb/src/xml4ngs/dist/ilmn2project.jar
   [delete] Deleting directory /home/lindenb/src/xml4ngs/tmp

download-ivy:
   (...)

install-ivy:

ivy.libs:
   (...)

dist/transformngsproject.jar:
    [mkdir] Created dir: /home/lindenb/src/xml4ngs/tmp
    [javac] Compiling 1 source file to /home/lindenb/src/xml4ngs/tmp
      [jar] Building jar: /home/lindenb/src/xml4ngs/dist/transformngsproject.jar
   [delete] Deleting directory /home/lindenb/src/xml4ngs/tmp

all:


\end{lstlisting}



\input{xsd.tex}

\section{Generating the XML project.}

\begin{lstlisting}[language=bash]
$java -jar ~/src/xml4ngs/dist/ilmn2project.jar \
	-f human_exome.xml \
	/dir/FASTQ > project.xml
\end{lstlisting}

you file then should look like this:
\begin{lstlisting}[language=XML,breaklines=true]
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://raw.github.com/lindenb/xml4ngs/master/src/main/resources/xsd/project.xsd">
  <properties>
    <property key="contaminations.reference.path">/commun/data/pubdb/contaminants/contaminants.fa</property>
    <property key="output.directory">../align</property>
    <property key="bwa.aln.options"> -q 15 -t 2 </property>
    (...)
  </properties>
  <lanes>
    <lane>1</lane>
    <lane>2</lane>
    <lane>3</lane>
    <lane>4</lane>
    <lane>5</lane>
    <lane>6</lane>
    <lane>7</lane>
    <lane>8</lane>
  </lanes>
  <sample name="FATHER">
    <sequences>
      <pair lane="3" split-index="2" sample-index="CGATGT">
        <fastq index="1" path="/dir/FASTQ/Sample_FATHER/FATHER_CGATGT_L003_R1_002.fastq.gz"/>
        <fastq index="2" path="/dir/FASTQ/Sample_FATHER/FATHER_CGATGT_L003_R2_002.fastq.gz"/>
      </pair>
      <pair lane="3" split-index="3" sample-index="CGATGT">
        <fastq index="1" path="/dir/FASTQ/Sample_FATHER/FATHER_CGATGT_L003_R1_003.fastq.gz"/>
        <fastq index="2" path="/dir/FASTQ/Sample_FATHER/FATHER_CGATGT_L003_R2_003.fastq.gz"/>
      </pair>
  (...)
   </sample>
   <sample name="CHILD">
    <sequences>
  (...)
      <pair lane="5" split-index="1" sample-index="GATCAG">
        <fastq index="1" path="/dir/FASTQ/Sample_CHILD/CHILD_GATCAG_L005_R1_001.fastq.gz"/>
        <fastq index="2" path="/dir/FASTQ/Sample_CHILD/CHILD_GATCAG_L005_R2_001.fastq.gz"/>
      </pair>
    </sequences>
  </sample>
</project>
\end{lstlisting}


\section{Generating the Makefile.}
\begin{lstlisting}[language=bash]
$ java -jar /path/to/xml4ngs/dist/transformngsproject.jar \
 	project.xml \
 	/path/to/xml4ngs/velocity/project2make.vm > Makefile
\end{lstlisting}

\section{Running the workflow on \sge{} using \qmake{}}
run \make{} with option '-n' (dry-run) to check if everything is fine.
\begin{lstlisting}[language=bash]
$ make -n  \
      variations.gatk.snpeff \
      variations.samtools.snpeff
\end{lstlisting}

you should also use \git{} to save the changes in your project.xml and your Makefile.
\begin{lstlisting}[language=bash]
$ git init
$ git add project.xml Makefile
$ git commit -m "my changes"
\end{lstlisting}
switch to TCSH, configure \sge{} and run \qmake{} using the option -j to specify the number of parallel jobs.
\begin{lstlisting}[language=bash]
$ tcsh
$ source /commun/sge_root/bird/common/settings.csh
$ setenv ARCH lx24-amd64
$ qmake -cwd -v PATH -l arch=lx24-amd64 -- -j 20 \
      variations.gatk.snpeff \
      variations.samtools.snpeff
\end{lstlisting}

\section{The velocity files.}
The previous version of the worklow merged all the aligned BAM after '\bwa{} sampe' and invoked all the steps in ordrer to generate an
annotated \VCF{} . This version is still available on git at : \href{http://github.com/todo}.
\begin{center}
\includegraphics[scale=0.3,angle=90]{workflow04.png}
\end{center}
The newer version split by chromosome  '\bwa{} sampe'. The workflow generates much more files but it is faster on \sge{}.
\begin{center}
\includegraphics[scale=0.3,angle=90]{workflow03.png}
\end{center}

\section{Property files.}
Hack: the XML property files used by 'ilmn2project.jar', specific for each type of analysis can be softly generated using \href{http://www.w3.org/TR/xinclude/}{x:include}.
\begin{center}
\includegraphics[scale=0.4]{workflow06.png}
\end{center}
For example 'template\_all.xml' could be:
\begin{lstlisting}[language=XML]
<?xml version="1.0"?>
<properties>
 <property key='bwa.aln.options'> -q 15 -t 2 </property>
 <property key='gatk.unified.genotyper.options'> -nt 20 </property>
 <property key='allele.calling.in.capture'>yes</property>
 <property key='one.vcf.per.sample'>yes</property>
 <property key="make.include">/commun/data/packages/tools.mk</property>
</properties>
\end{lstlisting}
and 'template\_human.xml' (it includes template\_all.xml )
\begin{lstlisting}[language=XML,escapeinside={!}{!}]
<?xml version="1.0"?>
<properties xmlns:xi="http://www.w3.org/2001/XInclude">
 !\redspan{ \textless{}xi:include href="template\_all.xml" xpointer="xpointer(/properties/property)" /\textgreater{} }!
 <property key='genome.reference.path'>/path/to/human_g1k_v37.fasta</property>
 <property key='known.sites'>/path/to/dbsnp_135.b37.vcf.gz</property>
</properties>
\end{lstlisting}
etc..\\
One can generate all the property files using \href{http://xmlsoft.org/xmllint.html}{xmllint}.
\begin{lstlisting}[language=bash]
$ xmllint --xinclude --format \
   templates/template_human_wholegenome.xml > \
   human_wholegenome.xml
\end{lstlisting} 
\section{Conclusion}
great
\section{Ackwnoledgements}
I'd like to thank the following persons:  Audrey Bihou√©e, Solena Le Scouarnec, Richard Redon, Raluca Teusan, Laetitia Duboscq-Bidot.

\end{document}
